### PR TITLE
feat: faucet devcon

### PR DESCRIPTION
**Pull request by ctrlc03**
Add next.js api handler to verify zupass proof - store the ticket id to a sqlite db to prevent re use of the same ticket. If proof is good then send some tokens to cover transaction fees.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -44,3 +44,6 @@ packages/interface/metamask/
 # typescript
 *.tsbuildinfo
 zkeys
+
+db.db 
+proofs
```
Comment (0xmad) [inline .gitignore:49]: Let's add new line at the end

File: packages/interface/package.json [hunk 2]
```diff
@@ -60,6 +60,8 @@
     "react-number-format": "^5.3.1",
     "react-use": "^17.5.0",
     "sonner": "^1.4.0",
+    "sqlite": "^5.1.1",
+    "sqlite3": "^5.1.7",
```
Comment (0xmad) [inline packages/interface/package.json:64]: Do we need these both packages?

File: packages/interface/src/components/JoinButton.tsx [hunk 3]
```diff
@@ -17,12 +17,28 @@ import type { EdDSAPublicKey } from "@pcd/eddsa-pcd";
...
+    // first we call the faucet to get funds
+    const response = await fetch("/api/faucet", {
+      method: "POST",
+      body: JSON.stringify({ address, pcd: zupassProof }),
+      headers: {
+        "Content-Type": "application/json",
+      },
+    });
+
+    if (response.status !== 204 && response.status !== 200) {
```
Comment (0xmad) [inline packages/interface/src/components/JoinButton.tsx:36]:
```suggestion
    if (!response.ok) {
```

File: packages/interface/src/components/JoinButton.tsx [hunk 4]
```diff
@@ -17,12 +17,28 @@ import type { EdDSAPublicKey } from "@pcd/eddsa-pcd";
...
-  const handleSignup = useCallback(() => onSignup(onError), [onSignup, onError]);
+  const handleSignup = useCallback(async () => {
+    // first we call the faucet to get funds
+    const response = await fetch("/api/faucet", {
+      method: "POST",
+      body: JSON.stringify({ address, pcd: zupassProof }),
+      headers: {
+        "Content-Type": "application/json",
+      },
+    });
```
Comment (0xmad) [inline packages/interface/src/components/JoinButton.tsx:34]: I think it should be moved to service layer.
@kittybest moved some logic recently to `server/api` but I don't remember what exactly.

File: packages/interface/src/pages/api/faucet.ts
Comment (0xmad) [inline packages/interface/src/pages/api/faucet.ts:1]: This one should not be in the pages (should be moved to server/api)

File: packages/interface/src/pages/api/faucet.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+ * @param req - The request object.
+ * @param res - The response object.
+ * @returns - The response object.
+ */
+export default async function handler(req: NextApiRequest, res: NextApiResponse): Promise<void> {
+  const db = await getDb();
+
+  try {
+    // we expect a JSON body with a pcd and an address
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
```
Comment (0xmad) [inline packages/interface/src/pages/api/faucet.ts:32]: You can use `as` cast here (or `as unknown as Type`)

File: packages/interface/src/pages/api/faucet.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      res.status(401).json("Invalid address");
+    }
+
+    if (!(await ZKEdDSAEventTicketPCDPackage.verify(pcd))) {
+      console.error(`[ERROR] ZK ticket PCD is not valid`);
+      res.status(200).json("Not valid");
+      return;
+    }
+
+    if (pcd.claim.signer[0] !== "044e711fd3a1792a825aa896104da5276bbe710fd9b59dddea1aaf8d84535aaf") {
```
Comment (0xmad) [inline packages/interface/src/pages/api/faucet.ts:54]: Let's use constant here

File: packages/interface/src/pages/api/faucet.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    if (alreadyUsed) {
+      res.status(204);
+      return;
+    }
+
+    if (!isAddress(address)) {
+      res.status(401).json("Invalid address");
+    }
+
+    if (!(await ZKEdDSAEventTicketPCDPackage.verify(pcd))) {
```
Comment (0xmad) [inline packages/interface/src/pages/api/faucet.ts:48]:
```suggestion
    const isValid = await ZKEdDSAEventTicketPCDPackage.verify(pcd);
    if (!isValid) {
```

File: packages/interface/src/pages/api/faucet.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+export default async function handler(req: NextApiRequest, res: NextApiResponse): Promise<void> {
+  const db = await getDb();
+
+  try {
+    // we expect a JSON body with a pcd and an address
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    const { address, pcd }: { address: string; pcd: PCD<ZKEdDSAEventTicketPCDClaim, Groth16Proof> } = req.body;
+
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    const alreadyUsed = await db.get("SELECT * FROM users WHERE pcd = ?", pcd.id);
```
Comment (0xmad) [inline packages/interface/src/pages/api/faucet.ts:36]: I think here we can use query in a more efficient way:
```
SELECT EXISTS(SELECT 1 FROM users WHERE pcd = ?);
```

File: packages/interface/src/pages/api/faucet.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+      return;
+    }
+
+    if (!isAddress(address)) {
+      res.status(401).json("Invalid address");
+    }
+
+    if (!(await ZKEdDSAEventTicketPCDPackage.verify(pcd))) {
+      console.error(`[ERROR] ZK ticket PCD is not valid`);
+      res.status(200).json("Not valid");
```
Comment (0xmad) [inline packages/interface/src/pages/api/faucet.ts:50]: Maybe we can use different status code for errors?

**Review by 0xmad — CHANGES_REQUESTED** [review]
@ctrlc03 thanks, just some changes requested

File: packages/interface/src/components/JoinButton.tsx (hunk 4 above)
Comment (kittybest) [reply packages/interface/src/components/JoinButton.tsx:34]: Currently, the `fetch` functions are called inside the `utils` folder, and for frontend components to call the api, we put the call inside `server/api` for the components to fetch by calling `api.faucet.getFund` for example.

**Comment by ctrlc03** [comment]
Closing as not needed

**Pull request closed without merging**